### ISSUE TITLE
Handling of content

### ISSUE THREAD

**Issue by neithernut** [matthiasbeyer/imag#120]
I do think @matthiasbeyer, @TheNeikos and I agree on that an entry may contain a chunk of textual content, maybe as a MIME container or something similar, as well as optionally refer to items outside the store via URIs. The rationale is that you can embed small datasets (e.g. appointments) in the store itself but refer to data which can not/should not be embedded into the store nicely (e.g. media files). Another possibility opening up would be to provide a "sync" address via an URI (e.g. to a webdav) for a piece of embedded data.
The current `entry` type defined in the store is not yet capable of serving any of these features. Things which have to be done are:
 * [x] define how URIs are provided in the header
 * [x] provide methods for retrieval of the URIs
 * [x] provide means to store the ~~MIME-type of the data~~ type of text embedded. ~~either by a similar declaration within the header or by using/writing a MIME-container for the `content`~~
 * [ ] clarify what the semantics are on some operations. When getting the content, for example, do we return an empty string if no data is embedded in the item file itself or do we try to fetch the data via one of the URIs?

**Comment by TheNeikos** [comment]
I am not sure how much you want to tie the Store to the Data. I don't think it is a good idea to go down that road.

There is one exception I am willing to make though: The header should have a special table called 'content' that if present tells the store that it doesn't care about the content on disk as it is only a reference. (And the rest of the data stays in the header anyway.)

An example:

### An entry generated by an `imag-deincylinder` module

```toml
---
[imag]

[imag.content]
url = "https://www.deincylinder.com/watch?v=randomid"

mime = "video/mp4"
file = "/home/neikos/Videos/deincylinder/winter-wrap-up.mp4"

[youtube]
rating = 9 # Out of 10
viewed = 958 # Times
---
```

So there are two things going on, `imag.content.url` is the official location of the resource, it is the only __required__ entry. The other two are module specific, but I think that we now have the chance for good defaults to increase interoperability. For example in my use cases I will mostly want to use it to save pictures/Videos/Music I see, classify them and download them neatly so that I can easily retrieve them offline later on (Should the URL expire, it be taken down or just no internet connection).

# Advantages 

You liberate the Store of a lot of heavy lifting that might not be needed in the end. The store itself doesn't need to know the MIME type for example.

# Disadvantages

Each module would need to separately parse and check the entries. In my opinion this makes for more code, but also gives a lot of flexibility regarding the structure of the external content. However, the need for a 'best practices' guide is very apparent to make sure that common types are not fragmented.

**Comment by matthiasbeyer** [comment]
@neithernut 

> I do think @matthiasbeyer, @TheNeikos and I agree on that an entry may contain a chunk of textual content, maybe as a MIME container or something similar, 

I'm not sure on this one. The `content`-Part of a file is meant for user-input only.

> as well as optionally refer to items outside the store via URIs. 

Yes.

> The rationale is that you can embed small datasets (e.g. appointments) in the store itself but refer to data which can not/should not be embedded into the store nicely (e.g. media files). 

Yes, though the "datasets" you're talking of would be in the header, not in the content part.

> Another possibility opening up would be to provide a "sync" address via an URI (e.g. to a webdav) for a piece of embedded data.

No. This is not the domain of `imag` yet. We _might_ do this at a point far in future, but it is not even included in my overall goal. WebDav syncing and such is domain of external programs, we should just aware of that files might be rewritten (updated by sync) or (re)moved while we're pointing to them.

> Things which have to be done are:
> * define how URIs are provided in the header

This is a point. I propose a `[links]` header section (which is generic, of course). Modules shall be free to place URIs there. This would be implemented in `libimagmodule`.

> provide methods for retrieval of the URIs

Would be implemented with `libimagmodule`. This lib can get some frontends for commandline-retrieval then.

> provide means to store the MIME-type of the data embedded, either by a similar declaration within the header or by using/writing a MIME-container for the content

I'm not completely sure about the whole MIME stuff you're talking about. I see that it makes sense for _some_ MIME types to have them. But we (especially me) want to keep the store entries _plain text_. So MIME types like `application/octet-stream` for example are just not our domain.

I agree, though, that we want content type identification. For example we want

* Markdown
* TexTile
* ReStructured Text

in the `content` Part of the files. This has nothing to do with MIME, though.

> clarify what the semantics are on some operations. When getting the content, for example, do we return an empty string if no data is embedded in the item file itself or do we try to fetch the data via one of the URIs?

This is true. But as it is domain of the modules itself (expect for the raw-content fetching out of the `content` part of the entries), it has to be resolved in the module libs/binaries and the respective interfaces.

---

@TheNeikos 

I'm not sure on your part here. We clearly want some way to identify external content in a generic way, and MIME makes sense here. Though I'm not sure whether to make these things _required_, as content still could be stored in the `content` part, making the `imag.content.url` variable obsolete in these cases. Maybe I misunderstood you here.

I like the general idea, it just has to be reworked a bit.

1. Don't name it `imag.content`, but just `content`. The `imag` section is _only_ for the very most generic information about the program (currently the version of the store library). We can argue about this and I could understand if you'd prefer the `imag.`-prefix, though.
1. The whole section is _optional_.
1. Call the variables `external.uri`, `external.mime` and `local.uri`, `local.mime`. (or `s/local/localcopy/`).
1. The `external.uri` is explicitely an URL, not a pointer to local content.
1. The `local` section _might_ contain more meta-information on the linked file, as it might get moved, etc. SHA-sums or something to be able to re-find it. I'm also not sure about this one, as MP3-tags change and change the SHA sum as well. We have to argue here.

**Comment by TheNeikos** [comment]
Could you put together an example or otherwise illustrate your ideas? It gets hard to follow when it is intersperse like that.

---

I think you misread/misunderstood what I meant. Basically if the `whatever.content` entry exists then it means that this file has external content and should be handled differently. External in this case means *not in the file*. It might just as well point to another store entry. If this entry exists, thus the `url` field is required. Not all the time.

> 1\. The imag section is only for the very most generic information about the program

Do we have a discussion about this somewhere? I see no reason for that to be necessarily true, and it would also allow a module name 'content' to easily use its rightfullly own namespace. 

> 3\. Call the variables external.uri, external.mime and local.uri, local.mime. (or s/local/localcopy/).

I don't think it is a good idea to have a dichotomy between local/external built into the store. It shouldn't matter to the store, and libraries are free to do what they see fit.

> 4\. The external.uri is explicitely an URL, not a pointer to local content.

[Please](https://en.wikipedia.org/wiki/File_URI_scheme) besides it makes it possible to put in more than just files.

**Comment by matthiasbeyer** [comment]
Yes, I misunderstood. Your points are completely legit!

---

Another idea, just popping up in my head:

Only allow links store-internally but providing a module which does external-linking. So one could simply link to an entry which then links to external content.

This is not fully thought through, though I think it might be a valuable path... ?

**Comment by neithernut** [comment]
> > The rationale is that you can embed small datasets (e.g. appointments) in the store itself but refer to data which can not/should not be embedded into the store nicely (e.g. media files).

> Yes, though the "datasets" you're talking of would be in the header, not in the content part.

So you would introduce your own format for appointments, instead of just using iCalendar?

> I agree, though, that we want content type identification. For example we want
>
>    Markdown
>    TexTile
>    ReStructured Text

Exactly. We may not have to store the mime-type in each entry for external content. But if you store data in the entry-file itself, you need a way to store the type of the content, even for text. Consider, for example, a viewer for items, which gives you a nice little display of your formatted text. How do you know what parser/formatter to use for displaying the content? Heuristics? Try to parse it a s MD, then fall back to something else? That's ugly.

> in the content Part of the files. This has nothing to do with MIME, though.

The point is: why would you come up with your own content type naming scheme if you can just use mime-types, like `text/plain` or `text/rtf`? I have to admit, though, that I did not yet stumble upon the mime-type for Markdown files.

----

I think I have to point out that I did _not_ request that `imag` gains an auto-downloader or something similar, but the way content which is external to the store is referred to should be defined globally and _not_ for each module. This way, one can write modules using the URIs and things "just work" because the tools stick to a common convention.
As @TheNeikos pointed out: content which is external to the store may well be local files. URIs allow this via the `file://` scheme, as I thought would be common knowledge.

**Comment by neithernut** [comment]
Wahrg, wrong button. Sry.

**Comment by matthiasbeyer** [comment]
> So you would introduce your own format for appointments, instead of just using iCalendar?

No. ICalendar files are external content and are linked into the store, not copied to it.

> How do you know what parser/formatter to use for displaying the content?

That's why I said we need an identifier for content in MD/TexTile/ReStructuredText.

> The point is: why would you come up with your own content type naming scheme if you can just use mime-types, like `text/plain` or `text/rtf`?

Because MIME is way to powerful. We have basically three content types, the said ones and plaintext.

> but the way content which is external to the store is referred to should be defined globally and not for each module

I completely agree here.

**Comment by neithernut** [comment]
Mh... I smell emerging incompatibilities...
But if you insist on those design choices...
I'll adjust the issue text.

**Comment by matthiasbeyer** [comment]
Am I right when saying that this is resolved by #122 ? May I close this issue?

**Comment by neithernut** [comment]
#122 is still not merged, so even if it resolves this issue, it should only be closed if #122 is merged.
The first two points are covered by that PR. As far as I can see the third and fourth point is still open.

**Comment by matthiasbeyer** [comment]
Point 3 is resolved. Content-type is domain of modules.

On point 4: Yes, you're basically right about this, though I don't think we should do things implicitely.

**Comment by neithernut** [comment]
I think the last point can also be marked as done if #136 is merged and/or the functionality it introduces regarding the semantics of content retrieval is properly documented.
